docs/syntax-if-statements-initializers.md

```diff
@@ -6,13 +6,13 @@ Introduce an initializer expression that declares and initializes variables in i
 
 # Motivation
 
-Initializers can improve code clarity and reduce scope pollution by allowing developers to declare variables in the conditions of if statements. ⁤⁤ Declaring variables at the point of use in if statements for the scope of the if statement's block simplifies code, leading to better readability and understanding of program logic. By limiting the scope of variables to the condition's pertaining block, the risk of unintended variable reuse and naming conflicts is reduced. ⁤
+Initializers can improve code clarity and reduce scope pollution by allowing developers to declare variables in the conditions of if statements. ⁤⁤ Declaring variables at the point of use in if statements for the scope of the if statement's blocks simplifies code, leading to better readability and understanding of program logic. By limiting the scope of variables to the if statement's blocks, the risk of unintended variable reuse and naming conflicts is reduced. ⁤
 
 The reduced scope pollution improves register space in extreme cases (or auto-generated code) where developers have many variables defined and have to work around register limits by reducing the register size. In some scenarios, especially on Roblox, initializing variables within if statements can lead to improved performance and reduced complexity by avoiding unnecessary calls by developers. ⁤ A common paradigm used by Roblox developers is to use `Instance:FindFirstChild` in their condition and then access it afterwards instead of keeping an existing variable around in the new scope, polluting the existing scope.
 
 # Design
 
-If statements with initializers must match (following the Luau grammar) `'if' 'local' bindinglist ['=' explist] 'then'` and `'local' bindinglist ['=' explist] where exp 'then'` syntax. The variables declared by the initializer are only available to the block of that condition and will be undefined to the `elseif` conditions and blocks as well as the `else` block; any code after the if statement won't have the variables defined either.
+If statements with initializers must match (following the Luau grammar) `'if' 'local' bindinglist ['=' explist] 'then'` and `'local' bindinglist ['=' explist] where exp 'then'` syntax. The variables declared by an initializer are only available to the if statement's blocks; any code after the if statement won't have the variables defined.
```

Comment:
This grammar has two issues: it lets you omit the expression, and it still allows for multiple bindings without a `where` clause. You probably want (taken from https://luau-lang.org/grammar):

```diff
  stat = varlist '=' explist |
      ...
      'repeat' block 'until' exp |
-     'if' exp 'then' block {'elseif' exp 'then' block} ['else' block] 'end' |
+     'if' cond 'then' block {'elseif' cond 'then' block} ['else' block] 'end' |
      'for' binding '=' exp ',' exp [',' exp] 'do' block 'end' |

- ifelseexp = 'if' exp 'then' exp {'elseif' exp 'then' exp} 'else' exp
+ ifelseexp = 'if' cond 'then' exp {'elseif' cond 'then' exp} 'else' exp

+ cond = 'local' binding '=' exp ['where' exp] |
+     'local' bindinglist '=' explist 'where' exp |
+     exp
```

I also noticed that my ask for adding initializers in `if` expressions earlier (reflected above) also permits this:

```lua
if local x = if local ok, res = pcall(f) where ok then res else nil where x == nil then
  -- ...
end
```

and I think we'd all want to live without seeing that.

Reply:
I'll update my grammar examples with your change, are you asking for if expressions with initializers too or are you against them?

Reply:
I think allowing initializers in `if` expressions is ok.

Reply:
I know that implementation complexity is sometimes completely ignored during design of a language, but I would suggest excluding extension of `if` expressions with this.
Reason being that current compiler model doesn't support local allocations inside expressions (all expression registers are temporary and can be clobbered).
I would understand if there was a great motivation for `if` expressions to rewrite large parts of the compiler, but seeing that this was more of an after-thought and nice-to-have, I suggest excluding them to a separate proposal (and I would probably ask for a compiler prototype in that).

Note that this will also be a reason why `do ... end` cannot become an expression so easily.